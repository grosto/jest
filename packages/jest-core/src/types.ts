/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {Context} from 'jest-runtime';
import {Test} from 'jest-runner';
import {Config, TestResult} from '@jest/types';

export type TestRunData = Array<{
  context: Context;
  matches: {
    allTests: number;
    tests: Array<Test>;
    total?: number;
    stats?: Stats;
  };
}>;

type TestPathCaseStats = Record<keyof (TestPathCases), number>;

export type TestPathCaseWithPathPatternStats = Record<
  keyof (TestPathCasesWithPathPattern),
  number
>;

export type Stats = TestPathCaseStats | TestPathCaseWithPathPatternStats;

export type TestPathCases = {
  roots: (path: Config.Path) => boolean;
  testMatch: (path: Config.Path) => boolean;
  testPathIgnorePatterns: (path: Config.Path) => boolean;
  testRegex: (path: Config.Path) => boolean;
};

export type TestPathCasesWithPathPattern = TestPathCases & {
  testPathPattern: (path: Config.Path) => boolean;
};

// TODO: Obtain this from @jest/reporters once its been migrated
export type ReporterOnStartOptions = {
  estimatedTime: number;
  showStatus: boolean;
};

export type Reporter = {
  onTestResult: (
    test: Test,
    testResult: TestResult.TestResult,
    aggregatedResult: TestResult.AggregatedResult,
  ) => Promise<void>;
  onRunStart: (
    results: TestResult.AggregatedResult,
    options: ReporterOnStartOptions,
  ) => Promise<void>;
  onTestStart: (test: Test) => Promise<void>;
  onRunComplete: (
    contexts: Set<Context>,
    results: TestResult.AggregatedResult,
  ) => Promise<void>;
  getLastError: () => Error;
};
