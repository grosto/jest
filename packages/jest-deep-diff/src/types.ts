/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {NewPlugin} from 'pretty-format/src/types';

// options
export type DeepDiffOptions = Omit<FormatOptions, 'plugins'> & {
  plugins?: Array<Plugin>;
};

export type FormatOptions = Partial<FormatOptionsNormalized>;

export type DiffOptionsColor = (arg: string) => string; // subset of Chalk type

export type FormatOptionsNormalized = {
  aAnnotation: string;
  aColor: DiffOptionsColor;
  aIndicator: string;
  bAnnotation: string;
  bColor: DiffOptionsColor;
  bIndicator: string;
  changeColor: DiffOptionsColor;
  changeLineTrailingSpaceColor: DiffOptionsColor;
  commonColor: DiffOptionsColor;
  commonIndicator: string;
  commonLineTrailingSpaceColor: DiffOptionsColor;
  contextLines: number;
  emptyFirstOrLastLinePlaceholder: string;
  expand: boolean;
  includeChangeCounts: boolean;
  omitAnnotationLines: boolean;
  patchColor: DiffOptionsColor;
  serialize: (val: unknown) => string;
  plugins: Array<FormatPlugin>;
};

// plugins
export interface Plugin {
  serialize: Serialize;
  test: PluginTest;
  diff: CustomDiff;
  format: CustomFormat;
}

export type PluginTest = (a: unknown) => boolean;

export interface FormatPlugin {
  test: PluginTest;
  format: CustomFormat;
}

export interface DiffPlugin {
  test: PluginTest;
  diff: CustomDiff;
}

type Serialize = NewPlugin['serialize'];
export interface SerializePlugin {
  test: PluginTest;
  serialize: Serialize;
}

// DiffObject
export type Path = unknown;

export enum Kind {
  EQUAL,
  UPDATED,
  INSERTED,
  DELETED,
  UNEQUAL_TYPE,
}
export interface DiffObject<T1 = unknown, T2 = T1> {
  kind: Kind;
  path?: Path;
  a: T1;
  b: T2;
  childDiffs?: Array<DiffObject>;
}

export type Memos = {
  a: Map<unknown, number>;
  b: Map<unknown, number>;
  position: number;
};

export type DiffFunc<T1 = unknown, T2 = T1> = (
  a: T1,
  b: T2,
  path: Path | undefined,
  memos?: Memos,
) => DiffObject<T1, T2>;

export type CustomDiff = (
  a: unknown,
  b: unknown,
  path: Path | undefined,
  memos: Memos | undefined,
  diff: DiffFunc,
) => DiffObject<unknown, unknown>;

// format

export type LineGenerationOptions = {
  plugins: Array<FormatPlugin>;
};

export type Format<T1 = unknown, T2 = T1> = (
  diff: DiffObject<T1, T2>,
  context: Context,
  options: LineGenerationOptions,
) => Array<Line>;

export type FormatComplexDiff<T1 = unknown, T2 = T1> = (
  diff: DiffObject<T1, T2>,
  context: Context,
  options: LineGenerationOptions,
  format: Format,
) => Array<Line>;

export type CustomFormat = FormatComplexDiff<unknown, unknown>;

export enum LineType {
  COMMON,
  INSERTED,
  DELETED,
}

export interface Context {
  indent?: string;
  prefix?: string;
  sufix?: string;
  skipSerialize?: boolean;
}

export interface Line {
  type: LineType;
  prefix: string;
  val: unknown;
  suffix: string;
  indent: string;
  skipSerialize?: boolean;
}
