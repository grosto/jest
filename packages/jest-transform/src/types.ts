/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {RawSourceMap} from 'source-map';
import {Config} from '@jest/types';

export type ShouldInstrumentOptions = Pick<
  Config.GlobalConfig,
  | 'collectCoverage'
  | 'collectCoverageFrom'
  | 'collectCoverageOnlyFrom'
  | 'coverageProvider'
> & {
  changedFiles?: Set<Config.Path>;
};

export type Options = ShouldInstrumentOptions &
  Partial<{
    isCoreModule: boolean;
    isInternalModule: boolean;
  }>;

// This is fixed in source-map@0.7.x, but we can't upgrade yet since it's async
interface FixedRawSourceMap extends Omit<RawSourceMap, 'version'> {
  version: number;
}

export type TransformedSource = {
  code: string;
  map?: FixedRawSourceMap | string | null;
};

export type TransformResult = {
  code: string;
  originalCode: string;
  mapCoverage: boolean;
  sourceMapPath: string | null;
};

export type TransformOptions = {
  instrument: boolean;
};

export type CacheKeyOptions = {
  config: Config.ProjectConfig;
  instrument: boolean;
  rootDir: string;
};

export interface Transformer {
  canInstrument?: boolean;
  createTransformer?: (options?: any) => Transformer;

  getCacheKey?: (
    fileData: string,
    filePath: Config.Path,
    configStr: string,
    options: CacheKeyOptions,
  ) => string;

  process: (
    sourceText: string,
    sourcePath: Config.Path,
    config: Config.ProjectConfig,
    options?: TransformOptions,
  ) => string | TransformedSource;
}
