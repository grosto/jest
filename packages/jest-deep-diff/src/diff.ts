import getType = require('jest-get-type');
import {DiffObject, DiffPlugin, Memos, Path} from './types';
import {wrapIfHasCircularity} from './complex/circularObjects';
import {createEqual, createUnequalType, createUpdated} from './diffObject';
import {diffObjects} from './complex/object';
import {diffArrays} from './complex/array';
import {diffMaps} from './complex/map';
import {
  diffBooleans,
  diffDates,
  diffErrors,
  diffFunctions,
  diffNumbers,
  diffRegExps,
  diffStrings,
} from './primitives';

// eslint-disable-next-line @typescript-eslint/ban-types
type PrimitiveWrapper = String | Number | Boolean;

function diff(
  a: unknown,
  b: unknown,
  path?: Path,
  memos?: Memos,
  customDiffs: Array<DiffPlugin> = [],
): DiffObject {
  if (Object.is(a, b)) return createEqual(a, b, path);

  const diffWithCustomDiffs = (
    a: unknown,
    b: unknown,
    path?: Path,
    memos?: Memos,
  ) => diff(a, b, path, memos, customDiffs);

  for (const customDiff of customDiffs) {
    if (customDiff.test(a) || customDiff.test(b)) {
      return customDiff.diff(a, b, path, memos, diffWithCustomDiffs);
    }
  }

  const aType = getType(a);

  const bType = getType(b);
  if (aType !== bType) {
    return createUnequalType(a, b, path);
  }

  if (a instanceof Error) {
    return diffErrors(a, b as Error, path);
  }

  switch (aType) {
    case 'string':
      return diffStrings(a as string, b as string, path);
    case 'number':
      return diffNumbers(a as number, b as number, path);
    case 'boolean':
      return diffBooleans(a as boolean, b as boolean, path);
    case 'date':
      return diffDates(a as Date, b as Date, path);
    case 'regexp':
      return diffRegExps(a as RegExp, b as RegExp, path);
    case 'function':
      return diffFunctions(a as Function, b as Function, path);
  }

  // adapted from node assert
  // Use memos to handle cycles.
  if (memos === undefined) {
    memos = {
      a: new Map(),
      b: new Map(),
      position: 0,
    };
  } else {
    // We prevent up to two map.has(x) calls by directly retrieving the value
    // and checking for undefined. The map can only contain numbers, so it is
    // safe to check for undefined only.
    const bMemoA = memos.a.get(a);
    const bMemoB = memos.b.get(b);
    if (bMemoA !== undefined || bMemoB !== undefined) {
      return bMemoA === bMemoB
        ? createEqual(
            wrapIfHasCircularity(a, bMemoA),
            wrapIfHasCircularity(b, bMemoB),
            path,
          )
        : createUpdated(
            wrapIfHasCircularity(a, bMemoA),
            wrapIfHasCircularity(b, bMemoB),
            path,
          );
    }
    memos.position++;
  }

  memos.a.set(a, memos.position);
  memos.b.set(b, memos.position);

  if (!Object.is(Object.getPrototypeOf(a), Object.getPrototypeOf(b))) {
    return createUnequalType(a, b, path);
  }

  // TODO: DOM Node

  if (a instanceof Number || a instanceof String || a instanceof Boolean) {
    return Object.is(a.valueOf(), (b as PrimitiveWrapper).valueOf())
      ? createEqual(a, b, path)
      : createUpdated(a, b, path);
  }

  const diffWithMemosAndCustomDiffs = (a: unknown, b: unknown, path: Path) =>
    diffWithCustomDiffs(a, b, path, memos);

  switch (aType) {
    case 'array':
      return diffArrays(
        a as Array<unknown>,
        b as Array<unknown>,
        path,
        diffWithMemosAndCustomDiffs,
      );
    case 'object':
      return diffObjects(
        a as Record<string | number | symbol, unknown>,
        b as Record<string | number | symbol, unknown>,
        path,
        diffWithMemosAndCustomDiffs,
      );
    case 'map':
      return diffMaps(
        a as Map<unknown, unknown>,
        b as Map<unknown, unknown>,
        path,
        diffWithMemosAndCustomDiffs,
      );
    // case 'set':
    //   return compareSets(a as Set<unknown>, b as Set<unknown>, path, deepDiff);
    default:
      throw new Error('oopsie ' + aType);
  }
}

export default diff;
