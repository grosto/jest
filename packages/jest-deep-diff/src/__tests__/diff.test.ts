import {stringify} from 'jest-matcher-utils';
import diff from '../diff';
import {DiffObject, Kind} from '../types';

describe('primitive values', () => {
  [
    [true, false],
    [1, 2],
    [0, -0],
    [0, Number.MIN_VALUE], // issues/7941
    [Number.MIN_VALUE, 0],
    ['banana', 'apple'],
    ['1\u{00A0}234,57\u{00A0}$', '1 234,57 $'], // issues/6881
    [
      'type TypeName<T> = T extends Function ? "function" : "object";',
      'type TypeName<T> = T extends Function\n? "function"\n: "object";',
    ],
  ].forEach(([a, b]) => {
    test(`${stringify(a)} is not equal to ${stringify(b)}`, () => {
      expect(diff(a, b).kind).toStrictEqual(Kind.UPDATED);
    });
  });

  [
    [true, true],
    [1, 1],
    [NaN, NaN],
    ['abc', 'abc'],
    ['banana', 'banana'],
  ].forEach(([a, b]) => {
    test(`${stringify(a)} is equal to ${stringify(b)}`, () => {
      expect(diff(a, b).kind).toStrictEqual(Kind.EQUAL);
    });
  });

  test('null and undefined are type unequal', () => {
    expect(diff(undefined, null)).toStrictEqual({
      a: undefined,
      b: null,
      kind: Kind.UNEQUAL_TYPE,
    });
  });
});

describe('primitive wrappers', () => {
  [
    /* eslint-disable no-new-wrappers */
    [new Number(0), new Number(1)],
    [new Boolean(false), new Boolean(true)],
    [new String('ab'), new String('abcd')],
    /* eslint-enable no-new-wrappers */
  ].forEach(([a, b]) => {
    test(`${stringify(a)} is not equal to ${stringify(b)}`, () => {
      const expected: DiffObject = {
        a,
        b,
        kind: Kind.UPDATED,
      };
      expect(diff(a, b)).toStrictEqual(expected);
    });
  });

  [
    /* eslint-disable no-new-wrappers */
    [new Number(2), new Number(2)],
    [new Boolean(false), new Boolean(false)],
    [new String('ab'), new String('ab')],
    /* eslint-enable no-new-wrappers */
  ].forEach(([a, b]) => {
    test(`${stringify(a)} is equal to ${stringify(b)}`, () => {
      const expected: DiffObject = {
        a,
        b,
        kind: Kind.EQUAL,
      };
      expect(diff(a, b)).toStrictEqual(expected);
    });
  });

  [
    /* eslint-disable no-new-wrappers */

    [0, new Number(0)],
    [new Number(0), 0],
    ['abc', new String('abc')],
    [new String('abc'), 'abc'],
    /* eslint-enable no-new-wrappers */
  ].forEach(([a, b]) => {
    test(`${stringify(a)} is not type equal to ${stringify(b)}`, () => {
      const expected: DiffObject = {
        a,
        b,
        kind: Kind.UNEQUAL_TYPE,
      };
      expect(diff(a, b, undefined)).toStrictEqual(expected);
    });
  });
});

describe('RegExp', () => {
  test('regexps with same source and same flags are equal', () => {
    const a = /abc/g;
    const b = /abc/g;

    const expected: DiffObject = {
      a,
      b,
      kind: Kind.EQUAL,
    };

    expect(diff(a, b)).toStrictEqual(expected);
  });

  test('regexps with difference source are not equal', () => {
    const a = /a/g;
    const b = /abc/g;

    const expected: DiffObject = {
      a,
      b,
      kind: Kind.UPDATED,
    };

    expect(diff(a, b)).toStrictEqual(expected);
  });

  test('regexps with different flags are not equal', () => {
    const a = /abc/gsy;
    const b = /abc/g;

    const expected: DiffObject = {
      a,
      b,
      kind: Kind.UPDATED,
    };

    expect(diff(a, b)).toStrictEqual(expected);
  });
});

describe('Objects', () => {
  test('marks updated field', () => {
    const a = {
      a: 1,
      b: 2,
    };

    const b = {
      a: 1,
      b: 3,
    };

    const expected: DiffObject = {
      a,
      b,

      childDiffs: [
        {
          a: a.a,
          b: b.a,
          kind: Kind.EQUAL,
          path: 'a',
        },
        {
          a: a.b,
          b: b.b,
          kind: Kind.UPDATED,
          path: 'b',
        },
      ],
      kind: Kind.UPDATED,
      path: undefined,
    };

    expect(diff(a, b)).toStrictEqual(expected);
  });

  test('marks deleted field', () => {
    const a = {
      a: 1,
      b: 2,
    };

    const b = {
      a: 1,
    };

    const expected: DiffObject = {
      a,
      b,
      childDiffs: [
        {
          a: a.a,
          b: b.a,
          kind: Kind.EQUAL,
          path: 'a',
        },
        {
          a: a.b,
          b: undefined,
          kind: Kind.DELETED,
          path: 'b',
        },
      ],
      kind: Kind.UPDATED,
      path: undefined,
    };

    expect(diff(a, b)).toStrictEqual(expected);
  });

  test('marks inserted field', () => {
    const a = {
      a: 1,
    };

    const b = {
      a: 1,
      b: 2,
    };

    const expected: DiffObject = {
      a,
      b,
      childDiffs: [
        {
          a: a.a,
          b: b.a,
          kind: Kind.EQUAL,
          path: 'a',
        },
        {
          a: undefined,
          b: b.b,
          kind: Kind.INSERTED,
          path: 'b',
        },
      ],
      kind: Kind.UPDATED,
      path: undefined,
    };

    expect(diff(a, b)).toStrictEqual(expected);
  });

  test('different constructors are not equal', () => {
    class A {}
    class B {}

    const a = new A();
    const b = new B();

    const expected: DiffObject = {
      a,
      b,
      kind: Kind.UNEQUAL_TYPE,
    };

    expect(diff(a, b)).toStrictEqual(expected);
  });

  test('non-enumerable members should be skipped during equal', () => {
    const a = {
      x: 3,
    };
    Object.defineProperty(a, 'test', {
      enumerable: false,
      value: 5,
    });

    const b = {
      x: 3,
    };

    const expected: DiffObject = {
      a,
      b,
      childDiffs: [
        {
          a: 3,
          b: 3,
          kind: 0,
          path: 'x',
        },
      ],
      kind: Kind.EQUAL,
      path: undefined,
    };

    expect(diff(a, b)).toStrictEqual(expected);
  });

  test('non-enumerable symbolic members should be skipped during equal', () => {
    const a = {
      x: 3,
    };
    const mySymbol = Symbol('test');
    Object.defineProperty(a, mySymbol, {
      enumerable: false,
      value: 5,
    });

    const b = {
      x: 3,
    };

    const expected: DiffObject = {
      a,
      b,
      childDiffs: [
        {
          a: 3,
          b: 3,
          kind: 0,
          path: 'x',
        },
      ],
      kind: Kind.EQUAL,
      path: undefined,
    };

    expect(diff(a, b)).toStrictEqual(expected);
  });

  test('correctly propagates "updated" to parent', () => {
    const a = {
      a: 1,
      b: {
        c: {
          d: 2,
        },
      },
    };

    const b = {
      a: 1,
      b: {
        c: {
          d: 1,
        },
      },
    };

    expect(diff(a, b)).toMatchSnapshot();
  });
});

describe('Arrays', () => {
  test('marks updated field', () => {
    const a = [1, 2];
    const b = [1, 3];

    const expected: DiffObject = {
      a,
      b,
      childDiffs: [
        {
          a: a[0],
          b: b[0],
          kind: Kind.EQUAL,
          path: 0,
        },
        {
          a: a[1],
          b: b[1],
          kind: Kind.UPDATED,
          path: 1,
        },
      ],
      kind: Kind.UPDATED,
      path: undefined,
    };

    expect(diff(a, b)).toStrictEqual(expected);
  });

  test('marks inserted fields', () => {
    const a = [1];
    const b = [1, 2];

    const expected: DiffObject = {
      a,
      b,
      childDiffs: [
        {
          a: a[0],
          b: b[0],
          kind: Kind.EQUAL,
          path: 0,
        },
        {
          a: a[1],
          b: b[1],
          kind: Kind.INSERTED,
          path: 1,
        },
      ],
      kind: Kind.UPDATED,
      path: undefined,
    };

    expect(diff(a, b)).toStrictEqual(expected);
  });

  test('marks deleted fields', () => {
    const a = [1, 2];
    const b = [1];

    const expected: DiffObject = {
      a,
      b,
      childDiffs: [
        {
          a: a[0],
          b: b[0],
          kind: Kind.EQUAL,
          path: 0,
        },
        {
          a: a[1],
          b: b[1],
          kind: Kind.DELETED,
          path: 1,
        },
      ],
      kind: Kind.UPDATED,
      path: undefined,
    };

    expect(diff(a, b)).toStrictEqual(expected);
  });
});

describe.skip('Maps', () => {
  test('marks updated field', () => {
    const a = new Map([
      ['a', 1],
      ['b', 3],
    ]);
    const b = new Map([
      ['a', 1],
      ['b', 2],
    ]);

    const expected: DiffObject = {
      a,
      b,

      childDiffs: [
        {
          a: a.get('b'),
          b: b.get('b'),
          kind: Kind.UPDATED,
          path: {
            a: 'b',
            b: 'b',
            kind: Kind.EQUAL,
          },
        },
        {
          a: a.get('a'),
          b: b.get('a'),
          kind: Kind.EQUAL,
        },
      ],
      kind: Kind.UPDATED,
      path: undefined,
    };

    expect(diff(a, b)).toStrictEqual(expected);
  });

  test('marks deleted field', () => {
    const a = new Map([
      ['b', 3],
      ['a', 1],
    ]);
    const b = new Map([['b', 3]]);

    const expected: DiffObject = {
      a,
      b,
      childDiffs: [
        {
          a: a.get('b'),
          b: undefined,
          kind: Kind.DELETED,
          path: 'b',
        },
        {
          a: a.get('a'),
          b: b.get('a'),
          kind: Kind.EQUAL,
          path: {
            a: 'a',
            b: 'a',
            kind: Kind.EQUAL,
          },
        },
      ],
      kind: Kind.UPDATED,
      path: undefined,
    };

    expect(diff(a, b)).toStrictEqual(expected);
  });

  test('marks inserted field', () => {
    const a = new Map([['b', 1]]);
    const b = new Map([
      ['b', 1],
      ['a', 3],
    ]);

    const expected: DiffObject = {
      a,
      b,
      childDiffs: [
        {
          a: undefined,
          b: b.get('a'),
          kind: Kind.INSERTED,
          path: 'a',
        },
        {
          a: a.get('b'),
          b: b.get('b'),
          kind: Kind.EQUAL,
          path: {
            a: 'b',
            b: 'b',
            kind: Kind.EQUAL,
          },
        },
      ],
      kind: Kind.UPDATED,
      path: undefined,
    };

    expect(diff(a, b)).toStrictEqual(expected);
  });
  test('checks deep equality of keys and values', () => {
    const a = new Map([[{a: 1}, 'abcd']]);
    const b = new Map([
      [{a: 1}, 'dsds'],
      [{a: 1}, 'dsassa'],
      [{a: 1}, 'abcd'],
    ]);

    const expected = {
      a,
      b,
      kind: Kind.EQUAL,
      path: undefined,
    };

    expect(diff(a, b)).toStrictEqual(expected);
  });
});

describe('Circular objects', () => {
  test('properties with the same circularity are equal', () => {
    const a = {};
    a.x = a;
    const b = {};
    b.x = b;
    expect(diff(a, b)).toMatchSnapshot();
    expect(diff(b, a)).toEqual(diff(a, b));

    const c = {};
    c.x = a;
    const d = {};
    d.x = b;
    expect(diff(c, d)).toMatchSnapshot();
    expect(diff(d, c)).toEqual(diff(c, d));
  });

  test('properties with different circularity are not equal', () => {
    const a = {};
    a.x = {y: a};
    const b = {};
    const bx = {};
    b.x = bx;
    bx.y = bx;

    expect(diff(a, b)).toMatchSnapshot();
    expect(diff(b, a).kind).toEqual(diff(a, b).kind);

    const c = {};
    c.x = a;
    const d = {};
    d.x = b;
    expect(diff(c, d)).toMatchSnapshot();
    expect(diff(c, d).kind).toEqual(diff(d, c).kind);
  });

  test('are not equal if circularity is not on the same property', () => {
    const a = {};
    const b = {};
    a.a = a;
    b.a = {};
    b.a.a = a;
    expect(diff(a, b)).toMatchSnapshot();
    expect(diff(b, a).kind).toEqual(diff(a, b).kind);

    const c = {};
    c.x = {x: c};
    const d = {};
    d.x = d;
    expect(diff(c, d)).toMatchSnapshot();
    expect(diff(c, d).kind).toEqual(diff(d, c).kind);
  });
});
