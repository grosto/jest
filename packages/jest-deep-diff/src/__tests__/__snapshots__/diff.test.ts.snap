// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Circular objects are not equal if circularity is not on the same property 1`] = `
Object {
  "a": Object {
    "a": [Circular],
  },
  "b": Object {
    "a": Object {
      "a": Object {
        "a": [Circular],
      },
    },
  },
  "childDiffs": Array [
    Object {
      "a": Object {
        "circularToDepth": 0,
        Symbol(circularVal): Object {
          "a": [Circular],
        },
      },
      "b": Object {
        "a": Object {
          "a": [Circular],
        },
      },
      "kind": 1,
      "path": "a",
    },
  ],
  "kind": 1,
  "path": undefined,
}
`;

exports[`Circular objects are not equal if circularity is not on the same property 2`] = `
Object {
  "a": Object {
    "x": Object {
      "x": [Circular],
    },
  },
  "b": Object {
    "x": [Circular],
  },
  "childDiffs": Array [
    Object {
      "a": Object {
        "x": Object {
          "x": [Circular],
        },
      },
      "b": Object {
        "circularToDepth": 0,
        Symbol(circularVal): Object {
          "x": [Circular],
        },
      },
      "kind": 1,
      "path": "x",
    },
  ],
  "kind": 1,
  "path": undefined,
}
`;

exports[`Circular objects properties with different circularity are not equal 1`] = `
Object {
  "a": Object {
    "x": Object {
      "y": [Circular],
    },
  },
  "b": Object {
    "x": Object {
      "y": [Circular],
    },
  },
  "childDiffs": Array [
    Object {
      "a": Object {
        "y": Object {
          "x": [Circular],
        },
      },
      "b": Object {
        "y": [Circular],
      },
      "childDiffs": Array [
        Object {
          "a": Object {
            "circularToDepth": 0,
            Symbol(circularVal): Object {
              "x": Object {
                "y": [Circular],
              },
            },
          },
          "b": Object {
            "circularToDepth": 1,
            Symbol(circularVal): Object {
              "y": [Circular],
            },
          },
          "kind": 1,
          "path": "y",
        },
      ],
      "kind": 1,
      "path": "x",
    },
  ],
  "kind": 1,
  "path": undefined,
}
`;

exports[`Circular objects properties with different circularity are not equal 2`] = `
Object {
  "a": Object {
    "x": Object {
      "x": Object {
        "y": [Circular],
      },
    },
  },
  "b": Object {
    "x": Object {
      "x": Object {
        "y": [Circular],
      },
    },
  },
  "childDiffs": Array [
    Object {
      "a": Object {
        "x": Object {
          "y": [Circular],
        },
      },
      "b": Object {
        "x": Object {
          "y": [Circular],
        },
      },
      "childDiffs": Array [
        Object {
          "a": Object {
            "y": Object {
              "x": [Circular],
            },
          },
          "b": Object {
            "y": [Circular],
          },
          "childDiffs": Array [
            Object {
              "a": Object {
                "circularToDepth": 1,
                Symbol(circularVal): Object {
                  "x": Object {
                    "y": [Circular],
                  },
                },
              },
              "b": Object {
                "circularToDepth": 2,
                Symbol(circularVal): Object {
                  "y": [Circular],
                },
              },
              "kind": 1,
              "path": "y",
            },
          ],
          "kind": 1,
          "path": "x",
        },
      ],
      "kind": 1,
      "path": "x",
    },
  ],
  "kind": 1,
  "path": undefined,
}
`;

exports[`Circular objects properties with the same circularity are equal 1`] = `
Object {
  "a": Object {
    "x": [Circular],
  },
  "b": Object {
    "x": [Circular],
  },
  "childDiffs": Array [
    Object {
      "a": Object {
        "circularToDepth": 0,
        Symbol(circularVal): Object {
          "x": [Circular],
        },
      },
      "b": Object {
        "circularToDepth": 0,
        Symbol(circularVal): Object {
          "x": [Circular],
        },
      },
      "kind": 0,
      "path": "x",
    },
  ],
  "kind": 0,
  "path": undefined,
}
`;

exports[`Circular objects properties with the same circularity are equal 2`] = `
Object {
  "a": Object {
    "x": Object {
      "x": [Circular],
    },
  },
  "b": Object {
    "x": Object {
      "x": [Circular],
    },
  },
  "childDiffs": Array [
    Object {
      "a": Object {
        "x": [Circular],
      },
      "b": Object {
        "x": [Circular],
      },
      "childDiffs": Array [
        Object {
          "a": Object {
            "circularToDepth": 1,
            Symbol(circularVal): Object {
              "x": [Circular],
            },
          },
          "b": Object {
            "circularToDepth": 1,
            Symbol(circularVal): Object {
              "x": [Circular],
            },
          },
          "kind": 0,
          "path": "x",
        },
      ],
      "kind": 0,
      "path": "x",
    },
  ],
  "kind": 0,
  "path": undefined,
}
`;

exports[`Objects correctly propagates "updated" to parent 1`] = `
Object {
  "a": Object {
    "a": 1,
    "b": Object {
      "c": Object {
        "d": 2,
      },
    },
  },
  "b": Object {
    "a": 1,
    "b": Object {
      "c": Object {
        "d": 1,
      },
    },
  },
  "childDiffs": Array [
    Object {
      "a": 1,
      "b": 1,
      "kind": 0,
      "path": "a",
    },
    Object {
      "a": Object {
        "c": Object {
          "d": 2,
        },
      },
      "b": Object {
        "c": Object {
          "d": 1,
        },
      },
      "childDiffs": Array [
        Object {
          "a": Object {
            "d": 2,
          },
          "b": Object {
            "d": 1,
          },
          "childDiffs": Array [
            Object {
              "a": 2,
              "b": 1,
              "kind": 1,
              "path": "d",
            },
          ],
          "kind": 1,
          "path": "c",
        },
      ],
      "kind": 1,
      "path": "b",
    },
  ],
  "kind": 1,
  "path": undefined,
}
`;
